python/cudf/cudf/core/buffer/cow_buffer.py

```diff
@@ -0,0 +1,215 @@
+        cls: Type[T], data: Any, *, exposed: bool = False
+    ) -> T:
+        """Create a Buffer from an object exposing `__cuda_array_interface__`.
+
+        No data is being copied.
+
+        Parameters
+        ----------
+        data : device-buffer-like
+            An object implementing the CUDA Array Interface.
+        exposed : bool, optional
+            Mark the buffer as zero copied.
+
+        Returns
+        -------
+        Buffer
+            Buffer representing the same device memory as `data`
+        """
+
+        # Bypass `__init__` and initialize attributes manually
+        ret = super()._from_device_memory(data)
+        ret._finalize_init()
+        ret._zero_copied = exposed
+        return ret
+
+    @classmethod
+    def _from_host_memory(cls: Type[T], data: Any) -> T:
+        ret = super()._from_host_memory(data)
+        ret._finalize_init()
+        return ret
+
+    @property
+    def _is_shared(self):
+        """
+        Return `True` if `self`'s memory is shared with other columns.
+        """
+        return len(self._instances) > 1
+
+    def get_ptr(self, mode: str = "write") -> int:
+        """Device pointer to the start of the buffer.
+
+        Parameters
+        ----------
+        mode : str, default 'write'
+            Supported values are {"read", "write"}
+            If "write", when weak-references exist, they
+            are unlinked and the data pointed to may be modified
+            by the caller. If "read", the data pointed to
+            must not be modified by the caller.
+            Failure to fulfill this contract will cause
+            incorrect behavior.
+
+        See Also
+        --------
+        Buffer.get_ptr
+        SpillableBuffer.get_ptr
+        """
+        if mode == "read":
+            return self._ptr
+        elif mode == "write":
+            self._unlink_shared_buffers()
+            return self._ptr
+        else:
+            raise ValueError(f"Incorrect mode passed : {mode}")
+
+    def _getitem(self, offset: int, size: int) -> Buffer:
+        """
+        Helper for `__getitem__`
+        """
+        return self._from_device_memory(
+            cuda_array_interface_wrapper(
+                ptr=self._ptr + offset, size=size, owner=self.owner
+            )
+        )
+
+    def copy(self, deep: bool = True):
+        """
+        Return a copy of Buffer.
+
+        Parameters
+        ----------
+        deep : bool, default True
+            If True, returns a deep-copy of the underlying Buffer data.
+            If False, returns a shallow-copy of the Buffer pointing to
+            the same underlying data.
+
+        Returns
+        -------
+        Buffer
+        """
+        if deep or self._zero_copied:
+            return self._from_device_memory(
+                rmm.DeviceBuffer(ptr=self._ptr, size=self.size)
+            )
+        else:
+            copied_buf = CopyOnWriteBuffer.__new__(CopyOnWriteBuffer)
```

Comment:
```suggestion
            cls = type(self)
            copied_buf = cls.__new__(cls)
```
?

In case someone inherits